### PR TITLE
Some optimisations for Microsoft.CSharp.RuntimeBinder.SymbolTable

### PR DESCRIPTION
**Pull request by JonHanna**
* Merge ClearCache() into ctor

* Switch in IsOperator

Don't keep hitting virtual property, and don't repeat checks.

* Only get name property once per method when hunting conversions.

* Make two never-used return types void.

* Add parameter conversions during AddNamesOnType rather than looping twice

* Don't create store results in AddNamesInInheritanceHierarchy

(not needed any more)

* Don't call GetRuntimeEvents unless we've processed a member.

If we won't use it, we don't need it.

* Only iterate through members once

* Add to _namesLoadedForEachType before further work. Skip if already done

* Avoid some double type checks

### CODE REVIEW

**Comment by JonHanna** [comment]
Linux x64 Release Build

**Comment by JonHanna** [comment]
Test Linux x64 Release Build

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs [hunk 1]
```diff
@@ -2088,78 +2067,72 @@ private void AddConversionsForOneType(Type type)
...
-
-                method.Name == SpecialNames.CLR_PreIncrement ||
-                method.Name == SpecialNames.CLR_PostIncrement ||
-                method.Name == SpecialNames.CLR_PreDecrement ||
-                method.Name == SpecialNames.CLR_PostDecrement);
+            if (method.IsSpecialName && method.IsStatic)
+            {
+                switch (method.Name)
+                {
+                    case SpecialNames.ImplicitConversion:
```
Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs]: Can you please comment on why this list was modified?

Comment (JonHanna) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs]: You mean why there are some items in the original `==` chain that aren't in the `switch`? Those were duplicate values, which C# doesn't allow in a `switch`.

Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs]: I see :)

**Pull request merged**